# **ETL**

## Objectives

* Write your notebook objective here, for example, "Fetch data from Kaggle and save as raw data", or "engineer features for modelling"

## Inputs

* Write down which data or information you need to run the notebook 

## Outputs

* Write here which files, code or artefacts you generate by the end of the notebook 

## Additional Comments

* If you have any additional comments that don't fit in the previous bullets, please state them here. 



---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'/Users/elliebrawn/Documents/vscode-projects/Hackathon_2026_Group_1/jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'/Users/elliebrawn/Documents/vscode-projects/Hackathon_2026_Group_1'

---

# Import Packages and Libraries

I order to read `HR-Employee-Attrition-raw.csv` and perform the ETL pipeline process, we need to install the following libraries and packages:

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style("whitegrid")

---

# Load the Dataset

The first thing we need to do is load the raw dataset to make sure it works.

In [5]:
# Load the raw dataset
df = pd.read_csv("datasets/raw-data/HR-Employee-Attrition-raw.csv")

# Display the first few rows of the dataset to make sure the data has loaded correctly
df.head()

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,...,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,...,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,...,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,...,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,...,4,80,1,6,3,3,2,2,2,2


---

# 1. Extract

In this first section, Extraction, we are looking to gain a baseline understanding of what our data looks like before we move on to cleaning and transforming.

We are using the acronym - **DISH** - to guide this process, where the letters stand for:

| Letter | Function / Attribute |
| ------ | ------- |
|**D**| `.describe`|
|**I**| `.info()`|
|**S**| `.shape`|
|**H**| `.head()`|

We will also use the following extraction steps:

| Function / Attribute |
| -----|
|`.dtypes`|
|`.isnull.sum()`|

## 1.1 Display the shape of the data

Using `.shape`, we are going to look at the shape (number of columns and rows)

In [6]:
# Display the shape of the dataset
print(f"Shape of the dataset: {df.shape}")

Shape of the dataset: (1470, 35)


The dataset contains 1470 rows and 35 columns.

## 1.2 Display the data types of each column

Now, we are going to inspect the data types for each column in our dataframe.

In [7]:
# Outline the data types of each column in the dataset
print(f"Data types of each column:\n{df.dtypes}")

Data types of each column:
Age                          int64
Attrition                   object
BusinessTravel              object
DailyRate                    int64
Department                  object
DistanceFromHome             int64
Education                    int64
EducationField              object
EmployeeCount                int64
EmployeeNumber               int64
EnvironmentSatisfaction      int64
Gender                      object
HourlyRate                   int64
JobInvolvement               int64
JobLevel                     int64
JobRole                     object
JobSatisfaction              int64
MaritalStatus               object
MonthlyIncome                int64
MonthlyRate                  int64
NumCompaniesWorked           int64
Over18                      object
OverTime                    object
PercentSalaryHike            int64
PerformanceRating            int64
RelationshipSatisfaction     int64
StandardHours                int64
StockOptionLevel            

## 1.3 Display a summary of all columns in the dataset

In [8]:
# Use .info() to get a concise summary of the dataset, including the number of non-null entries
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

This method builds on what we already saw using `.shape` and `.dtypes`. We can see that the `Non-Null` counts are the same as the number of rows in the dataset, meaning we can assume there are **no missing values**. 

We'll check this out using `isnull().sum()`.

## 1.4 Check for Missing Values in the Dataset

In [9]:
## Check for null values in the dataset
print(f"Number of null values in each column:\n{df.isnull().sum()}")

Number of null values in each column:
Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
Ye

As we already guessed would happen (given what we saw with the `Non Null` counts), there are no missing values in this dataset that will require us to handle them. 

## 1.5 Display Summary Statistics for the Dataset

In this section, we will be using `.describe()` in order to get summary statistics for all of the numerical columns in our dataset. This will be a good initial test to see if there are any glaringly obvious anomalies that we might need to deal with in our **Transform** section.

In [10]:
# Use .describe() for summary statistics of the numerical columns in the dataset
df.describe()

,Age,DailyRate,DistanceFromHome,Education,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,HourlyRate,JobInvolvement,JobLevel,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
count,1470.000000,1470.000000,1470.000000,1470.000000,1470.0,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,...,1470.000000,1470.0,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000,1470.000000
mean,36.923810,802.485714,9.192517,2.912925,1.0,1024.865306,2.721769,65.891156,2.729932,2.063946,...,2.712245,80.0,0.793878,11.279592,2.799320,2.761224,7.008163,4.229252,2.187755,4.123129
std,9.135373,403.509100,8.106864,1.024165,0.0,602.024335,1.093082,20.329428,0.711561,1.106940,...,1.081209,0.0,0.852077,7.780782,1.289271,0.706476,6.126525,3.623137,3.222430,3.568136
min,18.000000,102.000000,1.000000,1.000000,1.0,1.000000,1.000000,30.000000,1.000000,1.000000,...,1.000000,80.0,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000
25%,30.000000,465.000000,2.000000,2.000000,1.0,491.250000,2.000000,48.000000,2.000000,1.000000,...,2.000000,80.0,0.000000,6.000000,2.000000,2.000000,3.000000,2.000000,0.000000,2.000000
50%,36.000000,802.000000,7.000000,3.000000,1.0,1020.500000,3.000000,66.000000,3.000000,2.000000,...,3.000000,80.0,1.000000,10.000000,3.000000,3.000000,5.000000,3.000000,1.000000,3.000000
75%,43.000000,1157.000000,14.000000,4.000000,1.0,1555.750000,4.000000,83.750000,3.000000,3.000000,...,4.000000,80.0,1.000000,15.000000,3.000000,3.000000,9.000000,7.000000,3.000000,7.000000
max,60.000000,1499.000000,29.000000,5.000000,1.0,2068.000000,4.000000,100.000000,4.000000,5.000000,...,4.000000,80.0,3.000000,40.000000,6.000000,4.000000,40.000000,18.000000,15.000000,17.000000


Looking at this set of summary statistics, we have performed a quick lookover to see if we can spot any glaring anomalies. Obvious things to check:
* `Age`: is the minimum under 18? Are there any minors that we might need to address (given data protection laws)? **No, minim age is 18 and maximum age is 60, which feels like a suitable age range**.
* Minus Numbers: there should be no minus numbers in this set of statistics given what the columns relate to. **A quick look over says there are none**.

## 1.6 Check for duplicate values

We are going to check for duplicates now to see if any need to be dropped in the **Transform** section.

In [15]:
# Check for duplicate rows in the dataset
print(f"Number of duplicate rows in the dataset: {df.duplicated().sum()}")

Number of duplicate rows in the dataset: 0


Having performed `df.duplicated().sum()`, we can see that there are no duplicate rows that we will need to deal with.

---